controlplane-rs/src/gateway_controller.rs

```diff
@@ -0,0 +1,178 @@
+*/
+
+use futures::StreamExt;
+use std::{
+    ops::Sub,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use crate::*;
+use k8s_gateway_api::{Gateway, GatewayClass, GatewayClassSpec, GatewaySpec};
+use k8s_openapi::api::core::v1::Service;
+use kube::{
+    api::{Api, ListParams, Patch, PatchParams},
+    runtime::{controller::Action, watcher::Config, Controller},
+    Resource, ResourceExt,
+};
+
+use chrono::Utc;
+use gateway_utils::*;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1 as metav1;
+use tracing::*;
+
+pub async fn reconcile(gateway: Arc<Gateway>, ctx: Arc<Context>) -> Result<Action> {
+    let start = Instant::now();
+    let client = ctx.client.clone();
+    let name = gateway.name_any();
+    let ns = gateway.namespace().unwrap();
+
+    let gateway_api: Api<Gateway> = Api::namespaced(client.clone(), &ns);
+    let gateway_spec: GatewaySpec = gateway.spec.clone();
+    let mut gw = Gateway {
+        metadata: gateway.metadata.clone(),
+        spec: gateway.spec.clone(),
+        status: gateway.status.clone(),
+    };
+
+    let gateway_class_api = Api::<GatewayClass>::all(client.clone());
+    let gateway_class = gateway_class_api
+        .get(gateway_spec.gateway_class_name.as_str())
+        .await
+        .map_err(Error::KubeError)?;
+
+    let gateway_class_spec: GatewayClassSpec = gateway_class.spec.clone();
+    if gateway_class_spec.controller_name.as_str() != GATEWAY_CLASS_CONTROLLER_NAME {
+        return Ok(Action::requeue(Duration::from_secs(3600)));
+    }
+    debug!(
+        "found a supported GatewayClass: {:?}",
+        gateway_class.name_any()
+    );
+
+    set_listener_status(&mut gw);
+    let accepted_cond = get_accepted_condition(&mut gw);
+    set_condition(&mut gw, accepted_cond.clone());
+    if accepted_cond.status == "False".to_string() {
+        let programmed_cond = metav1::Condition {
+            last_transition_time: accepted_cond.last_transition_time.clone(),
+            observed_generation: accepted_cond.observed_generation,
+            type_: "Programmed".to_string(),
+            status: "False".to_string(),
+            message: accepted_cond.message.clone(),
+            reason: "Invalid".to_string(),
+        };
+        set_condition(&mut gw, programmed_cond);
+    }
+
+    if accepted_cond.status == String::from("False") {
+        patch_status(&gateway_api, name, gw.status.as_ref().unwrap()).await?;
+        return Err(Error::InvalidConfigError(accepted_cond.message));
+    }
+
+    let service_api: Api<Service> = Api::namespaced(client, &ns);
+    let services = service_api
+        .list(&ListParams::default().labels(&format!("{}={}", GATEWAY_SERVICE_LAEBL, name)))
+        .await
+        .map_err(Error::KubeError)?;
+    if services.items.len() > 1 {
+        return Err(Error::LoadBalancerError(
+            "found more than 1 Service for this Gateway; multiple services are not supported"
+                .to_string(),
+        ));
+    }
```

Comment:
I think for the first pass it's probably OK to leave this here, but we should ideally put a `TODO` comment on it with a link to an issue where we will follow up later, and then we should probably log the names of all the `Service` resources we found in case this ever _does actually happen_ the user has a hint about which `Service` resources are offending.

Reply:
is there any scenario where we would actually want to support multiple `Service`s? If a `Gateway` has N listeners, we can just define N ports on the `Service` itself (which is what we do atm).

Reply:
I'd have to think about it. I guess for now so we can move forward quickly I think it would be reasonable to just add a TODO here to re-evaluate, and answer that question.